resources/src/operation/mod.rs

```diff
@@ -0,0 +1,513 @@
+        }
+    }
+
+    /// Returns true if the operation is a manage data operation
+    pub fn is_manage_data(&self) -> bool {
+        match self.detail {
+            OperationDetail::ManageData(_) => true,
+            _ => false,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct IntermediateOperation {
+    id: i64,
+    paging_token: String,
+    #[serde(rename = "type")] operation_type: String,
+    type_i: u32,
+    account: Option<String>,
+    funder: Option<String>,
+    starting_balance: Option<Amount>,
+    from: Option<String>,
+    to: Option<String>,
+    asset_type: Option<String>,
+    asset_code: Option<String>,
+    asset_issuer: Option<String>,
+    source_asset_type: Option<String>,
+    source_asset_code: Option<String>,
+    source_asset_issuer: Option<String>,
+    amount: Option<Amount>,
+    source_amount: Option<Amount>,
+    source_max: Option<Amount>,
+    buying_asset_type: Option<String>,
+    buying_asset_code: Option<String>,
+    buying_asset_issuer: Option<String>,
+    selling_asset_type: Option<String>,
+    selling_asset_code: Option<String>,
+    selling_asset_issuer: Option<String>,
+    offer_id: Option<i64>,
+    #[serde(rename = "price_r")] price_ratio: Option<PriceRatio>,
+    price: Option<Amount>,
+    signer_key: Option<String>,
+    signer_weight: Option<u8>,
+    master_key_weight: Option<u8>,
+    low_threshold: Option<u32>,
+    med_threshold: Option<u32>,
+    high_threshold: Option<u32>,
+    home_domain: Option<String>,
+    set_flags: Option<Vec<u32>>,
+    set_flags_s: Option<Vec<String>>,
+    clear_flags: Option<Vec<u32>>,
+    clear_flags_s: Option<Vec<String>>,
+    trustor: Option<String>,
+    trustee: Option<String>,
+    limit: Option<Amount>,
+    into: Option<String>,
+    name: Option<String>,
+    value: Option<String>,
+}
+
+impl<'de> Deserialize<'de> for Operation {
+    fn deserialize<D>(d: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let rep: IntermediateOperation = IntermediateOperation::deserialize(d)?;
+        match &rep.operation_type[..] {
+            "create_account" => {
+                if rep.account.is_none() || rep.funder.is_none() || rep.starting_balance.is_none() {
+                    Err(de::Error::custom(
+                        "Missing required params for create account operation",
+                    ))?;
+                }
+                Ok(Operation {
+                    id: rep.id,
+                    paging_token: rep.paging_token,
+                    type_i: rep.type_i,
+                    detail: OperationDetail::CreateAccount(
+                        self::create_account::CreateAccount::new(
+                            rep.account.unwrap(),
+                            rep.funder.unwrap(),
+                            rep.starting_balance.unwrap(),
+                        ),
+                    ),
+                })
+            }
+            "payment" => {
```

Comment:
You can probably match on the struct itself and simplify this code so that you don't need to do a lot of `is_none` checks. Since this will destructure into the variables on the left side of the match, you won't need to `unwrap` anything.

```rust
match *rep {
    IntermediateOperation {
        operation_type: "payment",
        from: Some(from),
        to: Some(to),
        asset_type: Some(asset_type),
        amount: Some(amount),
        ..
    } => {
        /* Create a new one since you know the values are good */
    },
    // If this matches you know its bad
    IntermediateOperation { operation_type: "payment", .. } => Err(...)
}
```

Reply:
Some of the operations have the exact same signature including ManageOffer and CreatePassiveOffer, so, unfortunately, we will not be able to deterministically switch on operation signature.